### ISSUE TITLE
[BUG] Memory issues in 21cmFast

### ISSUE THREAD

**Issue by jordanflitter** [21cmfast/21cmFAST#337]
Hello,

I've encountered numerous memory issues that prevent me from completing the simulation. I should stress that in my project I work with a lot more redshift samples than in the standard 21cmFast, which is why I'm having these issues. Here's a list of the bugs that I've found so far, sorted by their severity:

1. In wrapper.py, in run_lightcone() and run_coeval(), every time a C-function is called (via initial_conditions(), perturb_field(), etc) the "write" argument is missing. As a result, boxes are saved in the cache even if the user explicitly specifies to not save anything in the cache with write=False.
2. In wrapper.py, in run_lightcone() and run_coeval(), perturb_field() is called in a different redshift loop than the main one (where  spin_temperature() is called) and the list "perturb_" grows in an uncontrolled manner. I believe this was done to support the old option of using the perturbed box with the lowest redshift through the entire calculation. However, this implementation can result a segmentation fault error when the size of the list is huge. I moved the call to perturb_field() within the main redshift loop and that fixed the segmentation fault error I had. 
3. In wrapper.py, in run_lightcone() and run_coeval(), when either spin_temperature() or ionize_box() are called, the cleanup argument logic implies that allocated memory is freed only at the last iteration. This logic appears to be deliberate (maybe for speeding up the calculation?), but I find it to be harmful when many redshift samples are needed. I have fixed this logic so memory is freed at each iteration, and the runtime seems to be not affected considerably by that change.
4. In BrightnessTemperatureBox.c, init_ps() is called at line 55, but free_ps() isn't called at the end of the calculation. In fact, the C-code in BrightnessTemperatureBox.c never uses anything that was initialized by init_ps(), and so that line should be removed.
5. In IonisationBox.c, init_ps() is called at line 112 but free_ps() isn't called at the end of the calculation. In addition, I found that the C-code in IonisationBox.c needs init_ps() only when mini-halos are turned on.
6. In SpinTemperatureBox.c, the variable "log10_Mcrit_LW_ave_list" is allocated (only when mini-halos are turned on), but never freed.
7. Finally, in wrapper.py, in run_lightcone() and run_coeval(), filenames of the calculated coeval boxes are appended to lists (again, in an uncontrolled manner), even if nothing was written to the cache (by fixing bug #1 in this list). 

Fixing all the above bugs seem to reduce the memory requirements of 21cmFast considerably. And yet, there is still something that keeps me from finishing my simulation. As far as I understand, when we get into the main redshift loop in either run_lightcone() or run_coeval(), the used memory by the program should remain on a fixed value, because new coeval boxes overwrite the previous ones at each iteration. However, this is certainly not what's happening, even when the above bugs are fixed.

To investigate what's causing this behavior, I wrote a script that calls to initial_conditions() multiples times, and prints the memory used by the returned InitialConditions object (I should note that I use the same random seed at each run). I should note that I found sys.getsizeof() to be unreliable for this task and I used instead actualsize(), which can be found in this nice [article](https://towardsdatascience.com/the-strange-size-of-python-objects-in-memory-ce87bdfbb97f). I was surprised to find that the memory used by each InitialConditions object was not constant, but rather growing with every call to initial_conditions(). Not only that, but the growth rate was not constant. I'm suspecting that it might have to do with the basic class "StructWrapper" that is used by the wrapper, but I am not sure. This is the point where I thought I should get some help from the github community :)

Thanks for anyone that can help me with that issue!

Jordan

**Comment by BradGreig** [comment]
Hi @jordanflitter, before being able to provide detailed answers to your question, can you please provide an example of what you are specifically aiming to compute (e.g. used variables, setup and version). That way we should be able to provide more useful information.

By memory issues, do you just mean you exhaust the available memory on your machine?

A brief response to some of your questions:

1. I know this used to be the case, but this was fixed a while ago. I just did a quick simulation using `run_lightcone` setting `write=False` and nothing was saved to the cache. This makes me wonder if you have an older version? Or you are attempting to do something else.

2. This has been identified in #221. This PR is an attempt to allow the user to estimate their memory requirements for their specific setup. It is not finalised yet, but you could try checking out that branch and it might give you some useful information for your setup.

The decision here was based on being able to purge the initial conditions, which individually require the most memory. But, if enough redshifts are required, this won't help for the identified reason. Likely a solution implementing both will be required (where the path taken depends on the user's setup).

3. Hmm, this hasn't caused an issue previously. I've always found the memory requirements to be fairly stable with repeated calls to `spin_temperature`. I'll have to have a think about this some further.

4 - 6. This is probably true, but the memory requirements here are essentially zero, essentially w.r.t `init_ps()`. So this should not be causing any problems.

7. Hmm, are you using the callback function? I think this should only be occurring if you are using the callback feature. I have never used this feature, so cannot speak for how it behaves with respect to memory.

Not sure if they are useful answers, but hopefully more assistance can be provided with more information on your desired setup

**Comment by steven-murray** [comment]
Thank you very much @jordanflitter for the detailed bug report. I agree with @BradGreig's comments, and I think a few of these can be fixed easily (and as he says, I'm pretty sure the first is already fixed on `master`). Making 21cmFAST memory-robust is important to us, so this will be a priority. Part of the problem is that there are different use-cases (as you well know) that require different code-paths to be optimal.

**Comment by jordanflitter** [comment]
Thanks @BradGreig and @steven-murray for showing interest in this issue. After more tests I did I found that the last issue (that prevented me from completing the simulation in my project) had to do with the modifications I did in my project, rather than another bug in 21cmFast :)

As for your questions, @BradGreig, I'm working with a modified code that is based on version 3.1.3. Because I saw that most of the lines that were causing me memory issues* are still found in the updated `master` code, I thought they could still lead to the same errors that I've encountered, though I understand that a fix to the first bug on my list was already implemented. 

*By memory issues I mean that the program has exhausted its available memory. For high resolutions I run the code on a Sun Grid Engine (SGE) system and many of the jobs were either killed/aborted by the system (in some occasions, a numpy ArrayMemoryError was raised). Before I fixed the 2nd bug on my list I used to get a segmentation fault error. I should say that when I work with a low resolution box on my laptop I've never had any of these problems. 

Before fixing the 3rd bug on my list, the memory usage by 21cmFast was increasing with each redshift step (here I used a poor resolution box of HII_DIM=25, no mini-halos, and I used the standard 84 redshift samples from z=35 to z=6. I made this plot with the wonderful tool [psrecord](https://github.com/astrofrog/psrecord). The first blue peak in this plot corresponds to the generation of the initial boxes)
 
![21cmFAST_public](https://github.com/21cmfast/21cmFAST/assets/131596329/c9a4e718-19a9-4557-90e3-fa7de80c5fdf)

Fixing the 3rd bug helps flattening the blue curve, but not entirely. Then, by fixing the 4th-6th bugs on my list I was able to bring the blue curve on a nearly** constant value. To be honest, I haven't checked what happens if the 7th bug on my list is fixed, I imagine it has a negligible effect on the total memory.

**Unfortunately, after making all these changes, the used memory by the 21cmFast is still not perfectly constant, and it gradually increases with each redshift step. In an attempt to figure out what may causing it, I wrote the following short script. This script returns different memory usage by the three different generated InitialConditions objects, which I find to be weird. I have to admit though that I don't have any prior experience with the actualsize(), although it seems to be doing a better job than sys.getsizeof(). 

```python
import sys
import gc
import numpy as np
import matplotlib.pyplot as plt
import py21cmfast as p21c

def actualsize(input_obj):
    memory_size = 0
    ids = set()
    objects = [input_obj]
    objects_memory = []
    while objects:
        new = []
        for obj in objects:
            if id(obj) not in ids:
                ids.add(id(obj))
                memory_size += sys.getsizeof(obj)
                new.append(obj)
                objects_memory.append(sys.getsizeof(obj))
        objects = gc.get_referents(*new)
    return memory_size, np.sort(objects_memory)

for i in np.arange(3):
    ICs = p21c.initial_conditions(user_params={'HII_DIM': 25 },
                                 regenerate=True,
                                 write=False,
                                 direc="_cache",
                                 random_seed=1)
    M, objects_memory = actualsize(ICs)
    print(f"Total memory of ICs is {M}")
    plt.plot(np.cumsum(objects_memory))
```

**Comment by BradGreig** [comment]
Hi @jordanflitter, excellent, good to know you were able to get your simulations complete!

The usage graph you've provided looks fairly similar to one I produce a while back (https://github.com/21cmfast/21cmFAST/pull/221#issuecomment-810746666). Apologies for the difficulty viewing it, it's run all the way through the calculation. However, between that version, your's and the current master there have been a few changes (mainly your 2nd point, which we are actively resolving).

Additionally, thanks for passing along the tools you've been using. They will be helpful (I was just scraping the information from top while the simulation was running).

We'll get to work when we can with respect to implementing a lot of these suggestions. I think the dominant one will be point 2. But we'll check and see how everything else is performing.

**Comment by jordanflitter** [comment]
Thanks @BradGreig. I'm glad that you found my comment helpful. Please let me know if you find anything interesting in that context :)

**Comment by jordanflitter** [comment]
Here’s my update of what’s going on in the code right now, with respect to the original list I made 3 years ago :)

1. **Saving boxes in the cache even if `write=False`**. This has been fixed. Today we have a nice decorator called `@single_field_func`. This decorator makes sure that if `write=False`, then the corresponding output object is not saved in the cache. In return, in `coeval.py`, every time a single field function is called, it includes as an argument `write=write.name_of_object`, where `write` on the RHS is the user’s `CacheConfig` object. **Status**: ✅
2. **Uncontrolled growth of `perturb_` list**. This actually still remains an issue. In `coeval.py`, in `_setup_ics_and_pfs_for_scrolling`, we still keep `append`ing `p` into `perturbed_field`, so if `all_redshifts` is very big (as well as `HII_DIM`), this could indeed lead to a huge memory consumption. Something similar happens in `evolve_halos` in `coeval.py`, where `halofield_list` grows uncontrollably. In this context, it is important to mention that there is a logic to `p.purge()` if `MINIMIZE_MEMORY` and `write.perturbed_field` are True, and there is a similar logic to `halos_desc.purge()` if `write.halo_catalog` is True. It is not clear to me why we use `MINIMIZE_MEMORY` only for the `perturbed_field`, but not for the `halo_catalog`, I suggest we examine it. Moreover, there could be some scenarios, like in computing the dark ages, where the need for many redshift samples dominants the memory budget compared to the size of `initial_conditions` (though the dark ages could be evaluated in a smart way that doesn’t require too many redshift samples, see #641), I suggest we also examine that as well. **Status**:❓
3. **Cleanup logic in last iteration**. We still have this cleanup logic at the last iteration, but only when `compute_spin_temperature` is called. I wonder why we have this logic only when this function is called, and not when other single-field functions are called. **Status**:❓
4. **Calling `init_ps` in `ComputeBrightnessTemp`, without `free_ps`**. This still remains as an issue on the main branch, however it is fixed in #624. **Status**: ✅
5. **Calling `init_ps` in `ComputeIonizedBox`, without `free_ps`**. This still remains as an issue on the main branch, however it is fixed in #624. **Status**: ✅
6. **`log10_Mcrit_LW_ave_list` is not freed in `SpinTemperatureBox.c`**. This has been fixed. This variable is now called `ave_log10_MturnLW` in the code. Unlike in v3, where it was dynamically allocated on the heap, it is now defined as a variable length array on the stack, which does not need to be freed. **Status**: ✅
7. **Uncontrolled growth of lists in `cache_files`**: This has been fixed. We no longer use `cache_files`. **Status**: ✅

@steven-murray do you have any comments about the above (2) and (3)? We could make them as sub-issues if you’d be interested.

Besides that, I should:
1. Test with [psrecord](https://github.com/astrofrog/psrecord) if the memory grows during an execution of `run_lightcone`.
2. Run my `actualsize` script from above and see if `initial_conditions` has a fixed size.

**Comment by steven-murray** [comment]
@jordanflitter thanks for the update! 

On 2   - so this doesn't accumulate memory if minimize memory is true? That is as intended then. The idea is that for a smaller run is faster to keep everything in memory. You should only purge if necessary, so we provide the flag. 

On 3 I will have to go look at the code again.

**Comment by steven-murray** [comment]
Hey @jordanflitter can you point me to the line number for issue (3)? I'm struggling to see what you're talking about

**Comment by jordanflitter** [comment]
Yes, it's this one: https://github.com/21cmfast/21cmFAST/blob/main/src/py21cmfast/drivers/coeval.py#L770.

**Comment by steven-murray** [comment]
Aha! I see. OK, I think the right strategy there would be to do what I did in https://github.com/21cmfast/21cmFAST/pull/624, and properly check the initialization compared to the inputs, and add the free function to the Python `atexit` registry. So that could be added a small sub-issue.

**Comment by jordanflitter** [comment]
I ran the following script with the great [psrecord](https://github.com/astrofrog/psrecord) tool.
```python
import py21cmfast as p21c

inputs = p21c.InputParameters.from_template(
    random_seed = 1, 
    name = ["latest","tiny"]
).with_logspaced_redshifts(
    zmin=6., 
    zmax=35., 
    zstep_factor=1.02,
)
lightconer = p21c.RectilinearLightconer.between_redshifts(
    min_redshift=inputs.node_redshifts[-1]+0.5,
    max_redshift=inputs.node_redshifts[0]-0.5,
    resolution=inputs.simulation_options.cell_size,
    cosmo=inputs.cosmo_params.cosmo,
)
lightcone = p21c.run_lightcone(
    inputs=inputs,
    lightconer=lightconer,
    progressbar=True,
    regenerate=True,
    write=p21c.CacheConfig.last_step_only(),
    include_dvdr_in_tau21=False,
)
```
To use `psrecord` (very to install, with `pip install psrecord`) one can run that script with
```
psrecord "python [path to script]" --plot [path to where to save plot]
```
I ran that script several times, with different templates and with different cache strategies. Here are the plots.

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/d747d3e7-b8f9-4aae-a78d-68b25fc09972" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/40721f13-88df-425f-ac27-3c90c8cb0a61" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/b74287f1-b7b3-4a7a-ae76-7740a92ad5ff" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/8f80b1b3-65e4-42ab-8a61-0f6a41c6f3ad" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/99087752-cd5c-442f-a1a1-27b6a1e46ad3" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/79e5e81c-870d-40ad-a110-4350b041d2fc" />

I'm writing my comments below for the above plots. As a general comment, I'll say that we can easily see the peak RAM memory at the beginning of the process, this is due to running `initial_conditions`.

1. **template=discrete, cache=last_step_only**. Memory usage remains stable as halo catalogs are computed and saved on disk. There is a weird peak when `evolve_halos` return. Then the memory keeps being stable until the program finishes its execution.
2. **template=discrete, cache=off**. Memory getting bigger when halos catalogs are computed and **not** saved on disk. This is expected and we actually throw a warning for this configuration. I do not understand the drop in RAM (which I think occurred when `evolve_halos` was in its last iterations) and the second increase towards the end.
3. **template=fixed-halos, cache=last_step_only**. Memory remains stable, I think the little drops in RAM are due to saving `HaloBoxes` in the cache.
4.  **template=fixed-halos, cache=off**. Memory gradually increases, I think it's due to keep having more and more `HaloBoxes` in the RAM, in order to compute `XraySourceBox` (at lower redshifts, `XraySourceBox` uses more `HaloBoxes`, in order to compute the filtered emissivity fields in retarded times). Note that we'll still have this design, even if we fix #552. I don't understand the sudden rise and drop in RAM, after `compute_initial_conditions` has returned.
5. **template=latest, cache=last_step_only**. I don't understand this behavior when `SOURCE_MODEL=E-INTEGRAL`...
6. **template=latest, cache=last_step_only**. Same as above, very strange behavior. **Note that the RAM increases with every iteration and eventually reaches a peak RAM that is greater even if we keep all the halo catalogs in RAM (with `template=discrete`)!!!**

**Comment by jordanflitter** [comment]
Also, it is important to note that the RAM behavior is **not** deterministic! I ran the 5th configuration (template=latest, cache=last_step_only) 3 times, and I got a different curve every time!

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/4ad013cd-e3f1-4244-b422-b628f75a9224" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/b9070e63-e2a1-44d8-a489-9f707073700e" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/ebe7b124-b652-4fc7-a751-8f967f748db6" />

**Comment by steven-murray** [comment]
Thanks @jordanflitter  --- this is potentially very helpful. It would be very cool if we could demarcate different sections of the code on these plots!

It might be worth including these as documentation somewhere.

The one I'm worried about is the last one (latest, with either cache off or llast_step_only). It is troubling that it is not deterministic (this could be somethig up with the tooling though?), and also that it seems to grow with the number of redshifts.

**Comment by jordanflitter** [comment]
Yeah, I agree, the tracked RAM curve with template="latest" doesn't look good. I think the curve is not deterministic because of something fundamental with our `OutputStruct`. If I run an updated version of my script from 3 years ago (the reason why I define `actualsize()` and not simply use `sys.getsizeof()` is because of this nice [article](https://towardsdatascience.com/the-strange-size-of-python-objects-in-memory-ce87bdfbb97f/)):
```python
import sys
import gc
import numpy as np
import py21cmfast as p21c

def actualsize(input_obj):
    memory_size = 0
    ids = set()
    objects = [input_obj]
    objects_memory = []
    while objects:
        new = []
        for obj in objects:
            if id(obj) not in ids:
                ids.add(id(obj))
                memory_size += sys.getsizeof(obj)
                new.append(obj)
                objects_memory.append(sys.getsizeof(obj))
        objects = gc.get_referents(*new)
    return memory_size, np.sort(objects_memory)

inputs = p21c.InputParameters(random_seed=1).evolve_input_structs(HII_DIM=25)
for i in range(10):
    ICs = p21c.compute_initial_conditions(
        inputs=inputs,
        regenerate=True,
        write=False,
    )
    M, objects_memory = actualsize(ICs)
    print(f"Total memory of ICs is {M}")
``` 
I get:
```
Total memory of ICs is 113349374
Total memory of ICs is 118156899
Total memory of ICs is 118156972
Total memory of ICs is 118157109
Total memory of ICs is 118157182
Total memory of ICs is 118157255
Total memory of ICs is 118157328
Total memory of ICs is 118157465
Total memory of ICs is 118157538
Total memory of ICs is 118157619
```
Notice that the total memory is increasing with every time `compute_initial_conditions` returns! (btw, if I reset the kernel of the notebook and rerun the script, I get similar numbers as above, but not exactly the same!)

**Comment by steven-murray** [comment]
It does keep increasing (which is worrying, and we should be able to detect this with valgrind), but also note that the increase after the second iteration is only ~100 bytes, which is not going to tank anyone's system very quickly :-P

**Comment by jordanflitter** [comment]
Regarding sub-issue 2 from my original list. After discussing with @steven-murray, we understand it makes sense that we don't use `MINIMIZE_MEMORY` in `evolve_halos`, since the halo catalogs consume a lot more memory than `perturbed_field`, namely we would always like to `purge` the halo catalog if it's written to disk. However, we would like to give the users more flexibility in the future in how the memory is handled throughout the simulation (for example, they might be interested in purging other `OutputStructs` that are written to disk, not just `perturbed_field` and `halo_catalog`).

Now that I'm looking again in the code of `evolve_halos`, I see that we `purge` `halos_desc`, but not `halos`. In other words, we `purge` the descendent halos from the **previous** iteration, but we do **not** `purge` the progenitor `halos` from the **current** iteration. @steven-murray doesn't it mean that we don't purge the halos from the last iteration? Shouldn't that be fixed?
Anyway, because we would like to replace the boolean `MINIMIZE_MEMORY` with a more flexible enum parameter, this sub-issue is not entirely fixed, and we shall fix that as part of a more up-to-date issue/PR. **Status**: ❓

Regarding sub-issue 3, I have checked that having `cleanup=True` at all iterations doesn't change the used RAM curve. Looking at my 21cmFirstCLASS code (based on v3), I am not sure why I had to eliminate the cleanup logic, because it seems like we still use the same cleanup logic (in both C and python). But anyway, this was somehow fixed. However, I will leave this sub-issue open (in a new up-to-date issue that I'll make shortly) because we could perhaps extend this logic to other single-field functions (I don't see a clear reason why we have this cleanup logic only in `compute_spin_temperature`). In addition, I see no reason why users should play with the `cleanup` boolean input to `run_coeval/lightcone`. **Status**: ❓

**Comment by jordanflitter** [comment]
I'm giving an update from the findings of the last week.

Apparently, the non-deterministic memory curves are a consequence of having the python's garbage collector on. When disabling it, the memory curve is the same in all runs, and it yields a peak memory which is much lower. This is what I get when I add `gc.disable()` to my script (in the plots below I work with template=latest, cache=off).

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/862afa97-1fc7-4b6c-84d6-e3e54df393d7" />

The memory accumulation at t > 17 sec is due to not purging the sub-structs of `prev_coeval` when `_redshift_loop_generator` finishes an iteration. When this is done, the memory curve becomes much more flat (but there is still some memory accumulation, though much less severe).

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/193201dc-783a-4735-9166-bb57f49b172c" />

These two issues have been solved in #690.

**Comment by jordanflitter** [comment]
Another update from the last days. Two bugs have been identified:

1. Dummy boxes (with `dummy=True`) were initialized and exposed to C. The dummy boxes have the default `inputs` with `HII_DIM=256` so they occupied quite a lot of memory that wasn't freed.
2. Initial boxes (with `initial=True`) were not purged at the end of their parent python function, even though they are exposed to C.

These two bugs caused the accumulation of memory in the above last plot at t=10-17 sec. When these issues are fixed, the curve is much flatter (but still not perfectly flat...)

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/663b20f4-01b2-45ed-88fc-22cd43639d18" />

Both bugs were fixed in #690.

**Comment by steven-murray** [comment]
Note that I think the fix to the `dummy` issue is the correct fix. The fix for `initial` boxes *works* but the long-term solution is to understand why on earth their memory is not being automatically freed. Why is there still a reference to these objects?!

**Comment by jordanflitter** [comment]
I think (but I am not sure) that the reference to these objects exists because they were exposed to C, which is probably why they are not automatically freed when they go out of scope.

I provide below the updated plots from psrecord, after #690 was merged. They look a lot better than before! Specifically, the final RAM is much lower, and there is no (severe) accumulation of memory. We can also identify in the plots where `z > Z_HEAT_MAX = 20` (for the `tiny` template), it happens at t > 17 sec (20 sec) for template=latest (fixed-halos). There are still though some unexplained phenomenona:

1. Why there is STILL some memory accumulation with template=latest?
2. What causes the big drop in RAM at t=100 sec in template=latest-discrete and cache=last_step_only? Is it because there are less halos at high redshifts and we purge the descendent halos? While it is excellent that the plateau between 200 < t < 400 sec is even lower than the RAM value we get with template=latest (i.e. with no discrete halos), this behavior is very strange.
3. Why the RAM curve for template=latest-discrete and cache=off does not grow monotonically? Why there are those sudden drops in RAM?

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/a1d33fa2-2e04-470a-94e1-de31aa5522bc" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/ef123ac6-efcc-4623-87a5-f72d5e23f0df" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/af0d83bb-7606-4f0a-9781-0d3a6616755a" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/34e14538-3d10-4726-957d-17a3fa66f09e" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/32d0306e-58e0-4fcb-a1d2-1b56965bd478" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/2836fbf8-eaae-44fb-92aa-e024b10c55b9" />

**Comment by steven-murray** [comment]
@jordanflitter if the objects are kept around simply because they are exposed to C, I am very confused. Python should garbage collect them once there are no references to them, and I don't think exposing them to C creates a reference.

The discrete halos case is indeed quite strange. IT would be awesome if we could figure out a way to mark events in the plot (like, which function is running when).

**Comment by jordanflitter** [comment]
Yes, @steven-murray I am as confused as you are, but I can say with confidence (based on the above plots) that purging structs that have been exposed to C is important, otherwise there could be a memory leak.

**Comment by jordanflitter** [comment]
I am closing this issue in favor to #691 (and also since #690 has been merged to the main branch).
